### PR TITLE
FEAT / Trainer: LOMO optimizer support

### LINKED ISSUE #29649: feat: LOMO support
### Feature request

Support for the LOMO training strategy, which allows full-weights training of 7B models on consumer 24GB cards like RTX 4090. 

ArXiv: https://arxiv.org/pdf/2306.09782.pdf https://arxiv.org/pdf/2310.10195.pdf
Implementation: https://github.com/OpenLMLab/LOMO (An installable package `lomo-optim` is also available here.)

### Motivation

This is another step towards democratizing LLM. And the variant AdaLomo performs on par with AdamW.

### Your contribution

I'm willing to submit a PR and modify the `lomo-optim` package if needed.
I may need some kind help for the integration on what I should do.

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes: https://github.com/huggingface/transformers/issues/29649

As requested by the community, this PR integrates LOMO optimizers into HF Trainer

https://github.com/OpenLMLab/LOMO 

I am facing some issues with respect to AdaLOMO, which seems to have DS as a hard requirement as in the optimizers init: https://github.com/OpenLMLab/LOMO/blob/85d8105c48cbd676dbf6915ee755461cd241da9b/lomo_optim/adalomo.py#L85 - so leaving this as a draft for now until I try to figure out this issue

cc @amyeroberts

### CODE REVIEW

**Comment by younesbelkada** [comment]
cc @muellerzr would love to have a first review! 🙏

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1363,6 +1369,24 @@ def optimizer_hook(param):
 
             if args.optim == OptimizerNames.GALORE_ADAFACTOR:
                 optimizer_kwargs.update({"scale_parameter": False, "relative_step": False})
+        elif args.optim in [OptimizerNames.LOMO, OptimizerNames.ADALOMO]:
+            if not is_lomo_available():
+                raise ImportError(
+                    "You need to install `galore_torch` in order to use GaLore optimizers"
+                    " install it with `pip install git+https://github.com/jiaweizzhao/GaLore`"
```
Comment (muellerzr) [inline src/transformers/trainer.py:1376]: This can be updated to use their pypi package, `galore-torch`

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1363,6 +1369,24 @@ def optimizer_hook(param):
...
+                )
+
+            if model is None:
+                raise ValueError("You need to pass a model in order to correctly initialize a LOMO optimizer.")
```
Comment (muellerzr) [inline src/transformers/trainer.py:1380]:
```suggestion
                raise ValueError("You need to pass a `model` in order to correctly initialize a LOMO optimizer.")
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2111,6 +2135,8 @@ def _inner_training_loop(
         self._globalstep_last_logged = self.state.global_step
         model.zero_grad()
         grad_norm: Optional[float] = None
+        # LOMO has a slightly different opitmizer API, see: https://github.com/OpenLMLab/LOMO/issues/73#issuecomment-2049612639
```
Comment (muellerzr) [inline src/transformers/trainer.py:2138]:
```suggestion
        # LOMO has a slightly different optimizer API, see: https://github.com/OpenLMLab/LOMO/issues/73#issuecomment-2049612639
```

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -2111,6 +2135,8 @@ def _inner_training_loop(
...
+        _is_lomo_optimizer = "Lomo" in self.optimizer.optimizer.__class__.__name__
```
Comment (muellerzr) [inline src/transformers/trainer.py:2139]: Can you check if we can peek at `self.optimizer.optimizer.__module__` maybe? (Or similar). Would be a bit more robust of a check than relying on namings

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -2261,8 +2287,9 @@ def _inner_training_loop(
                         else:
                             grad_norm = _grad_norm
 
-                    # Optimizer step
-                    self.optimizer.step()
+                    if not _is_lomo_optimizer:
```
Comment (muellerzr) [inline src/transformers/trainer.py:2290]: Let's keep the comment here, and explain why lomo doesn't need the step

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -3143,7 +3173,10 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             with amp.scale_loss(loss, self.optimizer) as scaled_loss:
                 scaled_loss.backward()
         else:
-            self.accelerator.backward(loss)
+            if not is_lomo_optimizer:
+                self.accelerator.backward(loss)
+            else:
+                self.optimizer.optimizer.fused_backward(loss, self._get_learning_rate())
```
Comment (muellerzr) [inline src/transformers/trainer.py:3179]: This makes me believe that we want to do this at the `Accelerate` level, to be quite honest. Simply because `accelerator.backward` handles grad accum dividing and a bunch of gradient scaling. 

Let's discuss this offline for a bit

**Review by muellerzr — COMMENTED** [review]
Thanks! This is a very interesting optimizer, so good to have this. Left some comments, I think it'd be best to downstream the support so accelerate can do `.backward()` easily and then bring it up to here

File: src/transformers/trainer.py (hunk 6 above)
Comment (younesbelkada) [reply src/transformers/trainer.py:3179]: yeah makes sense, happy to upstream that into accelerate to make things cleaner! ok will ping you offline

**Comment by younesbelkada** [comment]
thanks @muellerzr ! I offloaded most of the logic in https://github.com/huggingface/accelerate/pull/2695 - wdyt ? 🙏

**Review by muellerzr — APPROVED** [review]
Nice! Very well done. cc @LysandreJik for a final review

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -3201,12 +3246,24 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
 
                 The dictionary will be unpacked before being fed to the model. Most models expect the targets under the
                 argument `labels`. Check your model's documentation for all accepted arguments.
+            kwargs:
+                Additional key-word arguments to pass along for custom optimizers
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3250]: We should try to avoid adding in kwargs as much as possible, particularly when they're not needed. 

```suggestion
```

It's acceptable, although not ideal, to have the creation and preparation of kwargs within the method itself, but we really don't want to suddely allow uses to pass forward whatever they want to `self.accelerator.backward`

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -3187,7 +3232,7 @@ def autocast_smart_context_manager(self, cache_enabled: Optional[bool] = True):
 
         return ctx_manager
 
-    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> torch.Tensor:
+    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]], **kwargs) -> torch.Tensor:
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3235]:
```suggestion
    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> torch.Tensor:
```

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -3225,7 +3282,7 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             with amp.scale_loss(loss, self.optimizer) as scaled_loss:
                 scaled_loss.backward()
         else:
-            self.accelerator.backward(loss)
+            self.accelerator.backward(loss, **kwargs)
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3285]: What happens if we pass the learning rate through when lomo isn't being used?

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -3201,12 +3246,24 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
+        _is_lomo = False
+
+        if is_lomo_available():
+            from lomo_optim import AdaLomo, Lomo
+
+            _is_lomo = isinstance(_unwrap_optimizer(self.optimizer), (Lomo, AdaLomo))
+
+        # For LOMO optimizers you need to explicitly use the learnign rate
+        if _is_lomo:
+            kwargs["learning_rate"] = self._get_learning_rate()
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3266]: If we have the optimizer set, don't we also have `self._is_lomo_optimizer`?

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -251,6 +257,12 @@ def _is_peft_model(model):
     return False
 
 
+def _unwrap_optimizer(optimizer):
+    if isinstance(optimizer, AcceleratedOptimizer):
+        optimizer = optimizer.optimizer
+    return optimizer
```
Comment (amyeroberts) [inline src/transformers/trainer.py:263]: Is it guaranteed to only ever be one level of wrapping?

File: src/transformers/trainer.py [hunk 12]
```diff
@@ -3201,12 +3246,24 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
             `torch.Tensor`: The tensor with training loss on this batch.
         """
         model.train()
         inputs = self._prepare_inputs(inputs)
+        _is_lomo = False
+
+        if is_lomo_available():
+            from lomo_optim import AdaLomo, Lomo
+
+            _is_lomo = isinstance(_unwrap_optimizer(self.optimizer), (Lomo, AdaLomo))
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3262]: I'd move out the common logic to something like `_is_lomo_optimizer` from here and L1086, which handles importing Lomo and AdaLomo and unwrapping the optimizer

File: src/transformers/trainer.py [hunk 13]
```diff
@@ -3201,12 +3246,24 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
                 argument `labels`. Check your model's documentation for all accepted arguments.
+            kwargs:
+                Additional key-word arguments to pass along for custom optimizers
 
         Return:
             `torch.Tensor`: The tensor with training loss on this batch.
         """
         model.train()
         inputs = self._prepare_inputs(inputs)
+        _is_lomo = False
```
Comment (amyeroberts) [inline src/transformers/trainer.py:3257]: hmmmm..... needing to have this in the training_step is a good indication the abstractions here are leaky. Once we have the optimizer created, we shouldn't really need to know what type of optimizer it is in the rest of the code

**Review by amyeroberts — COMMENTED** [review]
Thanks for the work adding this! 

Some general comments about the structure. Similar to badam, the fact parts of the code need to know which optimizers are being used indicates we might be drawing the wrong boundaries around our abstractions.

File: src/transformers/trainer.py (hunk 11 above)
Comment (younesbelkada) [reply src/transformers/trainer.py:263]: Yes indeed ! https://github.com/huggingface/accelerate/blob/4ad4d28c49a9818e985ea12d66a89fe73fe73c87/src/accelerate/optimizer.py#L56

File: src/transformers/trainer.py (hunk 9 above)
Comment (younesbelkada) [reply src/transformers/trainer.py:3300]: It will break .. 😢  but we:
1- raise an error if users do not have the correct accelerate version with init-ing the trainer with lomo
2- pass `learning_rate` only if the optimizer is a lomo optimizer
3- removed `kwargs` in training step
So hopefully this should be safe enough 🙏

File: src/transformers/trainer.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/trainer.py]: yes indeed ! changed that

File: src/transformers/trainer.py (hunk 13 above)
Comment (younesbelkada) [reply src/transformers/trainer.py]: nice catch ..  i think it was an old code, now should be much cleaner !

**Review by amyeroberts — COMMENTED** [review]
Looks great! Thanks for all the work adding this feature + tests. 

I just have one comment about the `self._is_lomo_optimizer` flag. LMKWYT

File: src/transformers/trainer.py [hunk 14]
```diff
@@ -398,6 +410,7 @@ def __init__(
         self.hp_name = None
         self.deepspeed = None
         self.is_in_train = False
+        self._is_lomo_optimizer = False
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/trainer.py:413]: Makes sense, in that case what about just using `args.optim` as they can be used directly as enums: https://github.com/huggingface/transformers/blob/15c74a28294fe9082b81b24efe58df16fed79a9e/src/transformers/trainer.py#L1174 ? will commit something in that regard

File: tests/trainer/test_trainer.py [hunk 15]
```diff
@@ -1225,6 +1226,49 @@ def test_dataloader_without_dataset(self):
...
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # Trainer without inf/nan filter
+            args = TrainingArguments(tmpdir, learning_rate=1e-2, logging_steps=5, optim="lomo", max_steps=20)
+            trainer = Trainer(tiny_llama, args, train_dataset=train_dataset)
+
+            # Check this works
+            _ = trainer.train()
+
+        for name, param in tiny_llama.named_parameters():
+            self.assertFalse(torch.allclose(param, previous_params[name].to(param.device), rtol=1e-12, atol=1e-12))
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply tests/trainer/test_trainer.py:1249]: It is ok to put it higher, I decided to put it low so that even small changes would be captured by the test (sometimes higher tolerances would fail even though the weights are properly updated + with a high learning rate, so just to be on the safe zone)

File: src/transformers/trainer.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/trainer.py]: WDDYT of [9d547be](https://github.com/huggingface/transformers/pull/30178/commits/9d547be8dfe7f3494cf3c5e57564ca5bb1562fe6) ? With that I also removed the `_unwrap_optimizer` logic which makes some part of the code cleaner!

File: src/transformers/trainer.py [hunk 16]
```diff
@@ -239,6 +240,10 @@
     from accelerate.utils import DataLoaderConfiguration
 
 
+if is_lomo_available():
+    pass
+
+
```
Comment (younesbelkada) [inline src/transformers/trainer.py:246]:
```suggestion
```
This should be removed (in case the enum logic is validated)

File: src/transformers/trainer.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/trainer.py]: Looks great - thanks for iterating on this!

**Review by amyeroberts — APPROVED** [review]
Very nice - thanks for all the work adding this, including tests and iterating on the solution!

**Comment by younesbelkada** [comment]
Thanks so much for the extensive review and help @amyeroberts @muellerzr !
